src/Middleware/HttpLogging/src/W3CLoggerOptions.cs

```diff
@@ -106,6 +106,16 @@ public TimeSpan FlushInterval
         }
     }
 
+    /// <summary>
+    /// List of additional request header values to log.
+    /// The cookie header can contain authentication tokens,
+    /// or private information which may have regulatory concerns
+    /// under GDPR and other laws. Cookies should not be logged
+    /// unless logs are secure and access controlled
+    /// and the privacy impact assessed.
+    /// </summary>
+    public IReadOnlyList<string>? AdditionalRequestHeaders { get; set; }
```

Comment:
Thanks for doing this @dustinsoftware!

#41698 is still labeled `api-ready-for-review` which means that the public API changes have not been reviewed and are likely to change. There's even a chance the new API will be rejected entirely. I don't think that's very likely in this particular case, but it's something to be aware of when working on new API proposals that don't have the `api-approved` label yet.

I think this should probably follow the pattern set forth in [HttpLoggingOptions.RequestHeaders](https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/src/HttpLoggingOptions.cs#L18-L25) and use a get-only `ISet<string>` for these headers unless there's a problem with that approach.

```suggestion
    public ISet<string>? AdditionalRequestHeaders { get; }
```

Would it also make sense to also add a `ResponseHeaders` property to `W3CLoggerOptions` like we have in `HttpLoggingOptions`? Or would it be hard to distinguish request and response headers because of the W3C logging format?

Reply:
> Would it also make sense to also add a ResponseHeaders property to W3CLoggerOptions like we have in HttpLoggingOptions? Or would it be hard to distinguish request and response headers because of the W3C logging format?

Yeah I think that makes sense although that wasn't proposed in the original issue. To avoid this one getting blocked, I'd like to open a separate issue after this PR lands, unless they both need to be merged at the same time.

Reply:
> I think this should probably follow the pattern set forth in [HttpLoggingOptions.RequestHeaders](https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/src/HttpLoggingOptions.cs#L18-L25) and use a get-only ISet<string> for these headers unless there's a problem with that approach.

I didn't do that here since there is no ctor with args for W3CLoggingOptions, so the caller is relying on the public setter for the instance passed from the AddW3CLogging callback:

```
        services.AddW3CLogging(logging =>
        {
            // Log all W3C fields
            logging.LoggingFields = W3CLoggingFields.All;
            logging.AdditionalRequestHeaders = new[] {
                "x-forwarded-for",
                "x-client-ssl-protocol",
            }.ToHashSet();
        });
```

Reply:
> Yeah I think that makes sense although that wasn't proposed in the original issue.

Makes sense. We don't expect you to add any public API other than the approved `public ISet<string> AdditionalRequestHeaders { get; }`. I did mention this in the API review notes as a possible *future* enhancement.

> 2. Can we make this option more generic to support more field types like response headers?
> Maybe if we parsed prefixes like cs and sc, but that seems complicated. We could add a ResponseHeaders option in the future if we want.

`HttpLoggingOptions` doesn't have a constructor either. We expect the developer to use `.Clear()`/`.Add(string)`/`.Remove(string)` on the existing `ISet<string>` object to configure it:

https://github.com/dotnet/aspnetcore/blob/95374b37562dc97290f44b14f15196784550b8d1/src/Middleware/HttpLogging/test/HttpLoggingOptionsTests.cs#L57-L60

Reply:
I see. Updated :)